statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+    let (flush_tx, mut flush_rx) = mpsc::channel::<()>(1);
+
+    loop {
+        let (batch, shutdown) = construct_batch(
+            &mut rx,
+            Duration::from_millis(config.batch_creation_timeout_ms),
+            config.max_batch_size,
+        )
+        .await;
+
+        if !batch.is_empty() {
+            let permit = semaphore.clone().acquire_owned().await.unwrap();
+            stats
+                .requests_in_flight
+                .set((config.max_in_flight - semaphore.available_permits()) as f64);
+
+            let stats = stats.clone();
+            let retry_delay_ms = config.retry_delay_ms;
+            let client = client.clone();
+            let flush_tx = flush_tx.clone();
+            let trip = trip.clone();
+
+            tokio::spawn(async move {
+                let samples_in_batch = batch.len() as f64;
+                stats.requests_total.inc();
+                stats.samples_total.inc_by(samples_in_batch);
+
+                let write_request = construct_write_request(&stats, batch);
+                // N.B. According to the snap documentation, this compession should never fail for
+                // our use case.
+                let compressed_write_request = match compress_write_request(write_request) {
+                    Ok(c) => c,
+                    Err(e) => {
+                        stats.requests_compression_fail.inc();
+                        error!(
+                            "failed to compress prometheus WriteRequest, dropping samples: {:?}",
+                            e
+                        );
+                        return;
+                    }
+                };
+                stats
+                    .bytes_sent_total
+                    .inc_by(compressed_write_request.len() as f64);
+
+                // Attempt each request at least once
+                let mut res = client
+                    .send_write_request(compressed_write_request.clone())
+                    .await;
+                // And retry only if the tripwire has not yet been triggered.
+                while should_retry(&res) {
+                    sleep(Duration::from_millis(retry_delay_ms)).await;
+                    stats.requests_retry.inc();
+                    tokio::select! {
+                        new_res = client.send_write_request(compressed_write_request.clone()) => {
+                            res = new_res;
+                        },
+                        _ = trip.clone() => {
+                            break;
+                        },
+                    }
+                }
+
+                if let Err(e) = res {
+                    stats.requests_fail.inc();
+                    stats.samples_send_fail.inc_by(samples_in_batch);
+                    warn!(
+                        "prometheus remote write request failed, dropping samples: {:?}",
+                        e
+                    );
+                }
+
+                drop(permit);
```

Comment:
Minor nit: I'd recommend sticking a comment above these drops to indicate that they're here to force the move-into-closure. IME the rules around what gets closed over is a stumbling block for folks new to rust, so you can do them a favor when they read your code by tacking on a one-liner comment noting why the drops are explicit.

We'll need to do less of this as more of our team gets comfortable with rust, but for the time being IMO it's nice to try to make the gotchas more approachable.